src/docs_website/src/file_checker.zig

```diff
@@ -57,47 +63,243 @@ fn validate_dir(arena: std.mem.Allocator, path: []const u8) !void {
+
+    var link_iterator = LinkIterator.init(html);
+    errdefer log.err("[link checker] error in {s}:{}", .{
+        context.dir.realpathAlloc(context.arena, context.path) catch unreachable,
+        link_iterator.line_number,
+    });
+
+    while (link_iterator.next()) |link| {
+        try check_link(context, link);
+    }
+}
+
+fn check_link(context: FileValidationContext, link: Link) !void {
+    // Check schema.
+    {
+        if (std.mem.startsWith(u8, link.base, "mailto:")) {
+            return; // Ignore.
+        }
+
+        if (std.mem.startsWith(u8, link.base, "https://")) {
+            return check_external_link(context.arena, link);
+        }
+
+        if (std.mem.startsWith(u8, link.base, "http://")) {
+            if (http_exceptions.has(link.base)) {
+                return check_external_link(context.arena, link);
+            }
+
+            log.err("found insecure link: '{s}'", .{link.base});
+            return error.InsecureLink;
+        }
+    }
+
+    // Locate local link target.
+    var target = link.base;
+    const is_absolute = target.len > 0 and target[0] == '/';
+    if (is_absolute) {
+        target = target[1..];
+    } else if (std.fs.path.dirname(context.path)) |dirname| {
+        target = try std.fs.path.join(context.arena, &.{ dirname, target });
+    }
+
+    const is_directory = std.fs.path.extension(target).len == 0;
+    if (is_directory) {
+        target = try std.fs.path.join(context.arena, &.{ target, "index.html" });
+    }
+
+    if (!try path_exists(context.dir, target)) {
+        log.err("link target not found: '{s}'", .{target});
+        return error.TargetNotFound;
+    }
+
+    if (link.fragment) |anchor| {
+        try check_anchor(context, target, anchor);
+    }
+}
+
+fn check_external_link(arena: std.mem.Allocator, link: Link) !void {
+    if (!check_external_links) return;
+    if (https_exceptions.has(link.base)) return;
+
+    errdefer |err| log.err("got {} while checking external link '{s}'", .{ err, link.base });
+
+    log.info("checking external link '{s}'", .{link.base});
+
+    var client = std.http.Client{ .allocator = arena };
+    defer client.deinit();
+
+    const uri = try std.Uri.parse(link.base);
+    var header_buffer: [512 * 1024]u8 = undefined;
+    var request = try client.open(.GET, uri, .{ .server_header_buffer = &header_buffer });
+    defer request.deinit();
+
+    try request.send();
+    try request.finish();
+    try request.wait();
+
+    if (request.response.status != std.http.Status.ok) {
+        return error.WrongStatusResponse;
+    }
+}
+
+fn check_anchor(context: FileValidationContext, target_path: []const u8, anchor: []const u8) !void {
+    assert(std.mem.endsWith(u8, target_path, ".html"));
+
+    const html = try read_file_cached(context.arena, context.dir, target_path);
+    const needle = try std.mem.concat(context.arena, u8, &.{ "id=\"", anchor, "\"" });
+    if (std.ascii.indexOfIgnoreCase(html, needle) == null) {
```

Comment:
Shall we make this a case-sensitive comparison? I'd rather we fix our anchors to be exact than make the validator more lax!